In this part of the project we load and explore the data.

In [122]:
import pandas as pd 
import numpy as np

In [123]:
contract_df = pd.read_csv('./datasets/final_provider/contract.csv')
internet_df = pd.read_csv('./datasets/final_provider/internet.csv')
personal_df = pd.read_csv('./datasets/final_provider/personal.csv')
phone_df = pd.read_csv('./datasets/final_provider/phone.csv')

In [124]:
print('contract_df shape:', contract_df.shape)
print('contract_df:')
display(contract_df.head())
print('contract_df description:')
display(contract_df.describe())

contract_df shape: (7043, 8)
contract_df:


customerID   BeginDate              EndDate            Type  \
0  7590-VHVEG  2020-01-01                   No  Month-to-month   
1  5575-GNVDE  2017-04-01                   No        One year   
2  3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW  2016-05-01                   No        One year   
4  9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0              Yes           Electronic check           29.85        29.85  
1               No               Mailed check           56.95       1889.5  
2              Yes               Mailed check           53.85       108.15  
3               No  Bank transfer (automatic)           42.30      1840.75  
4              Yes           Electronic check           70.70       151.65

contract_df description:


MonthlyCharges
count     7043.000000
mean        64.761692
std         30.090047
min         18.250000
25%         35.500000
50%         70.350000
75%         89.850000
max        118.750000

In [125]:
print('internet_df shape:', internet_df.shape)
print('internet_df:')
display(internet_df.head())
print('internet_df description:')
display(internet_df.describe())

internet_df shape: (5517, 8)
internet_df:


customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0  7590-VHVEG             DSL             No          Yes               No   
1  5575-GNVDE             DSL            Yes           No              Yes   
2  3668-QPYBK             DSL            Yes          Yes               No   
3  7795-CFOCW             DSL            Yes           No              Yes   
4  9237-HQITU     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies  
0          No          No              No  
1          No          No              No  
2          No          No              No  
3         Yes          No              No  
4          No          No              No

internet_df description:


customerID InternetService OnlineSecurity OnlineBackup  \
count         5517            5517           5517         5517   
unique        5517               2              2            2   
top     7590-VHVEG     Fiber optic             No           No   
freq             1            3096           3498         3088   

       DeviceProtection TechSupport StreamingTV StreamingMovies  
count              5517        5517        5517            5517  
unique                2           2           2               2  
top                  No          No          No              No  
freq               3095        3473        2810            2785

In [126]:
print('personal_df shape:', personal_df.shape)
print('personal_df:')
display(personal_df.head())
print('personal_df description:')
display(personal_df.describe())

personal_df shape: (7043, 5)
personal_df:


customerID  gender  SeniorCitizen Partner Dependents
0  7590-VHVEG  Female              0     Yes         No
1  5575-GNVDE    Male              0      No         No
2  3668-QPYBK    Male              0      No         No
3  7795-CFOCW    Male              0      No         No
4  9237-HQITU  Female              0      No         No

personal_df description:


SeniorCitizen
count    7043.000000
mean        0.162147
std         0.368612
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000

In [127]:
print('phone_df shape:', phone_df.shape)
print('phone_df:')
display(phone_df.head())
print('phone_df description:')
display(phone_df.describe())

phone_df shape: (6361, 2)
phone_df:


customerID MultipleLines
0  5575-GNVDE            No
1  3668-QPYBK            No
2  9237-HQITU            No
3  9305-CDSKC           Yes
4  1452-KIOVK           Yes

phone_df description:


customerID MultipleLines
count         6361          6361
unique        6361             2
top     5575-GNVDE            No
freq             1          3390

After loading the data we take a look at the shapes of the data and what columns there are. We can see that there are a few data frames that dont match the shape of the rest. Later on we will correct this. 

In [128]:
df = pd.merge(contract_df, phone_df, on='customerID', how='left')
df = pd.merge(df, personal_df, on='customerID', how='left')
df = pd.merge(df, internet_df, on='customerID', how='left', validate='one_to_one')

df.head()

customerID   BeginDate              EndDate            Type  \
0  7590-VHVEG  2020-01-01                   No  Month-to-month   
1  5575-GNVDE  2017-04-01                   No        One year   
2  3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW  2016-05-01                   No        One year   
4  9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

  MultipleLines  gender  SeniorCitizen Partner Dependents InternetService  \
0           NaN  Female              0     Yes         No             DSL   
1            No    Male              0      No         No             DSL   
2            No    Male              0      No         No             DSL   
3           NaN    Male              0      No         No             DSL   
4            No  Female              0      No         No     Fiber optic   

  OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV  \
0             No          Yes               No          No          No   
1            Yes           No              Yes          No          No   
2            Yes          Yes               No          No          No   
3            Yes           No              Yes         Yes          No   
4             No           No               No          No          No   

  StreamingMovies  
0              No  
1              No  
2              No  
3              No  
4              No

Here we merge all the dataframes together and can begin work on improving our dataset for our machine learning task.

In [129]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
 8   MultipleLines     6361 non-null   object 
 9   gender            7043 non-null   object 
 10  SeniorCitizen     7043 non-null   int64  
 11  Partner           7043 non-null   object 
 12  Dependents        7043 non-null   object 
 13  InternetService   5517 non-null   object 
 14  OnlineSecurity    5517 non-null   object 
 15  OnlineBackup      5517 non-null   object 
 16  DeviceProtection  5517 non-null   object 


In [130]:
df[['BeginDate', 'EndDate']].head(5)

BeginDate              EndDate
0  2020-01-01                   No
1  2017-04-01                   No
2  2019-10-01  2019-12-01 00:00:00
3  2016-05-01                   No
4  2019-09-01  2019-11-01 00:00:00

In [131]:
df['BeginDate'] = pd.to_datetime(df['BeginDate'], format='%Y-%m-%d')
df['EndDate'] = pd.to_datetime(df['EndDate'], format='%Y-%m-%d %H:%M:%S', errors='coerce')
df[['BeginDate', 'EndDate']].head(5)

BeginDate    EndDate
0 2020-01-01        NaT
1 2017-04-01        NaT
2 2019-10-01 2019-12-01
3 2016-05-01        NaT
4 2019-09-01 2019-11-01

Above we correctly parse the date columns into dates. Not all end dates are avalable so we fill them in with 'not a time'.

In [132]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   customerID        7043 non-null   object        
 1   BeginDate         7043 non-null   datetime64[ns]
 2   EndDate           1869 non-null   datetime64[ns]
 3   Type              7043 non-null   object        
 4   PaperlessBilling  7043 non-null   object        
 5   PaymentMethod     7043 non-null   object        
 6   MonthlyCharges    7043 non-null   float64       
 7   TotalCharges      7043 non-null   object        
 8   MultipleLines     6361 non-null   object        
 9   gender            7043 non-null   object        
 10  SeniorCitizen     7043 non-null   int64         
 11  Partner           7043 non-null   object        
 12  Dependents        7043 non-null   object        
 13  InternetService   5517 non-null   object        
 14  OnlineSecurity    5517 n

In [133]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

The 'TotalCharges' column is currently an object so we will need to convert this to numerical for scaling. 

In [134]:
columns = ['MultipleLines', 'Partner', 'Dependents', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'PaperlessBilling']

df[columns] = df[columns].fillna('No')
df[columns] = df[columns].replace({'No': 0, 'Yes': 1})

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   customerID        7043 non-null   object        
 1   BeginDate         7043 non-null   datetime64[ns]
 2   EndDate           1869 non-null   datetime64[ns]
 3   Type              7043 non-null   object        
 4   PaperlessBilling  7043 non-null   int64         
 5   PaymentMethod     7043 non-null   object        
 6   MonthlyCharges    7043 non-null   float64       
 7   TotalCharges      7032 non-null   float64       
 8   MultipleLines     7043 non-null   int64         
 9   gender            7043 non-null   object        
 10  SeniorCitizen     7043 non-null   int64         
 11  Partner           7043 non-null   int64         
 12  Dependents        7043 non-null   int64         
 13  InternetService   5517 non-null   object        
 14  OnlineSecurity    7043 n

/var/folders/vt/tzztnb6n5td1dmz2kc5stqqr0000gn/T/ipykernel_54979/3332100295.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[columns] = df[columns].replace({'No': 0, 'Yes': 1})


Here we choose to fill in missing data for our binary columns. This effectively encodes them and will make it easer for our machine learning development. 

In [135]:
df['InternetService'] = df['InternetService'].fillna('None')

In [136]:
df = df.dropna(subset=['TotalCharges'])

In [137]:
duplicates = df.duplicated(subset='customerID', keep=False)
duplicates.value_counts()

False    7032
Name: count, dtype: int64

In the above cells we fill in 'InternetService' with None. We also drop missing 'TotalCharges'. Finally we check for duplicated customerID which there are none. 

## Questions
- Is it safe to fill in missing data or should rows that are missing data be dropped? 
- I dropped total charges, is that correct?
- Should the missing data in the InternetService column be filled in or dropped?

## Plan
The plan is to build a model that can accurately predict the churn for telecom operator Interconnect. I believe that this dataset lends itself well to a Classifier model. To start I will encode the categorical data such as columns 'PaymentMethod' and 'Type'. Then I will scale the numerical data in columns like 'MonthlyCharges' and 'TotalCharges'. After the data has been encoded I can start experimenting with Classifier algorithms. I am planning on experimenting with the following classifiers: 'DecisionTreeClassifier', 'LogisticRegression', and 'RandomForestClassifier'. After comparing the results I should be able to select a model that works for Interconnect.

In [138]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, roc_auc_score, classification_report, confusion_matrix
from sklearn.utils import shuffle

In [139]:
df['Churn'] = np.where(df['EndDate'].isnull(), 0, 1)
df.head()

customerID  BeginDate    EndDate            Type  PaperlessBilling  \
0  7590-VHVEG 2020-01-01        NaT  Month-to-month                 1   
1  5575-GNVDE 2017-04-01        NaT        One year                 0   
2  3668-QPYBK 2019-10-01 2019-12-01  Month-to-month                 1   
3  7795-CFOCW 2016-05-01        NaT        One year                 0   
4  9237-HQITU 2019-09-01 2019-11-01  Month-to-month                 1   

               PaymentMethod  MonthlyCharges  TotalCharges  MultipleLines  \
0           Electronic check           29.85         29.85              0   
1               Mailed check           56.95       1889.50              0   
2               Mailed check           53.85        108.15              0   
3  Bank transfer (automatic)           42.30       1840.75              0   
4           Electronic check           70.70        151.65              0   

   gender  SeniorCitizen  Partner  Dependents InternetService  OnlineSecurity  \
0  Female              0        1           0             DSL               0   
1    Male              0        0           0             DSL               1   
2    Male              0        0           0             DSL               1   
3    Male              0        0           0             DSL               1   
4  Female              0        0           0     Fiber optic               0   

   OnlineBackup  DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
0             1                 0            0            0                0   
1             0                 1            0            0                0   
2             1                 0            0            0                0   
3             0                 1            1            0                0   
4             0                 0            0            0                0   

   Churn  
0      0  
1      0  
2      1  
3      0  
4      1

Above we assume if the end date is set then the customer has churned. We store this value in a new 'Churn' column and this will become our target for our ML algorithm. Below we one-hot encode the data for several columns. These are categorical columns and one-hot encoding will help the machine learning libraries form better predictions on our data. 

In [140]:
df = pd.get_dummies(df, columns=['InternetService', 'Type', 'PaymentMethod', 'gender'], drop_first=True)
pd.set_option('display.max_columns', None)  
print(df.head())

   customerID  BeginDate    EndDate  PaperlessBilling  MonthlyCharges  \
0  7590-VHVEG 2020-01-01        NaT                 1           29.85   
1  5575-GNVDE 2017-04-01        NaT                 0           56.95   
2  3668-QPYBK 2019-10-01 2019-12-01                 1           53.85   
3  7795-CFOCW 2016-05-01        NaT                 0           42.30   
4  9237-HQITU 2019-09-01 2019-11-01                 1           70.70   

   TotalCharges  MultipleLines  SeniorCitizen  Partner  Dependents  \
0         29.85              0              0        1           0   
1       1889.50              0              0        0           0   
2        108.15              0              0        0           0   
3       1840.75              0              0        0           0   
4        151.65              0              0        0           0   

   OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  StreamingTV  \
0               0             1                 0            

Next we split our data into a test, train, and validation set. 

In [141]:
target = df['Churn']
features = df.drop(columns=['customerID', 'Churn', 'BeginDate', 'EndDate'], axis=1)

features_train, features_temp, target_train, target_temp = train_test_split(features, target, test_size=0.25, random_state=12345)

features_valid, features_test, target_valid, target_test = train_test_split(features_temp, target_temp, test_size=0.5, random_state=12345)

After that we scale our numerical columns. Scaling is important for 'MonthlyCharges' and 'TotalCharges' columns because the large number can alter the performance of several models we will use. By standardizing the range of the columns it provides a clearer signal to learn on.

In [142]:
numeric_cols = ['MonthlyCharges', 'TotalCharges']

scaler = StandardScaler()

features_train = features_train.copy()
features_valid = features_valid.copy()
features_test = features_test.copy()

features_train[numeric_cols] = scaler.fit_transform(features_train[numeric_cols])
features_valid[numeric_cols] = scaler.transform(features_valid[numeric_cols])
features_test[numeric_cols] = scaler.transform(features_test[numeric_cols])

display(features_train.head())

PaperlessBilling  MonthlyCharges  TotalCharges  MultipleLines  \
5495                 1        0.296184     -0.604484              0   
2009                 1        0.475876      1.507270              1   
1800                 1       -0.702104     -0.992243              0   
5630                 1        1.281161      0.743794              1   
6295                 0       -1.489087     -0.560101              0   

      SeniorCitizen  Partner  Dependents  OnlineSecurity  OnlineBackup  \
5495              0        0           0               0             0   
2009              0        1           1               1             1   
1800              0        0           0               0             0   
5630              0        0           0               0             1   
6295              0        1           1               0             0   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
5495                 0            1            0                0   
2009                 1            1            1                0   
1800                 0            0            0                0   
5630                 1            0            1                1   
6295                 0            0            0                0   

      InternetService_Fiber optic  InternetService_None  Type_One year  \
5495                         True                 False          False   
2009                        False                 False          False   
1800                        False                 False          False   
5630                         True                 False          False   
6295                        False                  True          False   

      Type_Two year  PaymentMethod_Credit card (automatic)  \
5495          False                                  False   
2009           True                                  False   
1800          False                                  False   
5630          False                                  False   
6295           True                                  False   

      PaymentMethod_Electronic check  PaymentMethod_Mailed check  gender_Male  
5495                            True                       False         True  
2009                           False                       False         True  
1800                           False                       False         True  
5630                            True                       False        False  
6295                           False                       False        False

## First Model

The first model we will try is the DecisionTreeClassifier model. This model is great for classification and regression. It creates simple rules for the data features. 

In [143]:
best_depth = 0
best_f1 = 0
best_roc_auc = 0
for depth in range(1,10):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train)
    predicted_valid = model.predict(features_valid)
    probabilities_valid = model.predict_proba(features_test)
    probabilities_one_valid = probabilities_valid[:, 1]

    if f1_score(target_valid, predicted_valid) > best_f1:
        best_f1 = f1_score(target_valid, predicted_valid)
        best_depth = depth
        best_roc_auc = roc_auc_score(target_valid, probabilities_one_valid)

print(f"Best max_depth = {best_depth} f1 score {best_f1} roc_auc {best_roc_auc}")

Best max_depth = 6 f1 score 0.5440806045340051 roc_auc 0.5486429291232785


In [144]:
model = DecisionTreeClassifier(random_state=12345, max_depth=6)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
print(classification_report(target_valid, predicted_valid))
probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
roc_auc = roc_auc_score(target_valid, probabilities_one_valid)
print(f"ROC AUC: {roc_auc}")

              precision    recall  f1-score   support

           0       0.83      0.91      0.87       650
           1       0.64      0.47      0.54       229

    accuracy                           0.79       879
   macro avg       0.74      0.69      0.71       879
weighted avg       0.78      0.79      0.78       879

ROC AUC: 0.8264057776284851


In [145]:
model = DecisionTreeClassifier(random_state=12345, max_depth=6, class_weight='balanced')
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
print(classification_report(target_valid, predicted_valid))
probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
roc_auc = roc_auc_score(target_valid, probabilities_one_valid)
print(f"ROC AUC: {roc_auc}")

              precision    recall  f1-score   support

           0       0.92      0.70      0.79       650
           1       0.49      0.83      0.62       229

    accuracy                           0.73       879
   macro avg       0.71      0.76      0.71       879
weighted avg       0.81      0.73      0.75       879

ROC AUC: 0.842311051394021


We have some good performance but we will try tuning the model even more. Below we are checking the balance of classes. As we can see a class imabalance exists. Below we will define functions to correct imbalance.

In [146]:
# CONFUSION MATRIX
result = confusion_matrix(target_valid, predicted_valid)
print(result)

[[454 196]
 [ 39 190]]


In [147]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)

    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345
    )

    return features_upsampled, target_upsampled

def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)]
        + [features_ones]
    )
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)]
        + [target_ones]
    )

    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345
    )

    return features_downsampled, target_downsampled

In [148]:
model = DecisionTreeClassifier(random_state=12345, max_depth=6, class_weight='balanced')

best_f1 = 0
best_repeat = 0
best_roc_auc = 0
for repeat in range(1,11):
    features_upsampled, target_upsampled = upsample(
        features_train, target_train, repeat
    )
    
    model.fit(features_upsampled, target_upsampled)
    predicted_valid = model.predict(features_valid)

    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    f1 = f1_score(target_valid, predicted_valid)
    if f1 > best_f1:
        best_f1 = f1
        best_repeat = repeat
        best_roc_auc = roc_auc_score(target_valid, probabilities_one_valid)
print('Repeat:', best_repeat)
print('F1:', best_f1)
print('ROC AUC:', best_roc_auc)

Repeat: 1
F1: 0.6178861788617886
ROC AUC: 0.842311051394021


In [149]:
best_f1 = 0
best_repeat = 0
best_roc_auc = 0
for repeat in range(1,6):
    features_downsampled, target_downsampled = downsample(
        features_train, target_train, (repeat/10)
    )

    model.fit(features_downsampled, target_downsampled)
    predicted_valid = model.predict(features_valid)

    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    f1 = f1_score(target_valid, predicted_valid)
    if f1 > best_f1:
        best_f1 = f1
        best_repeat = repeat
        best_roc_auc = roc_auc_score(target_valid, probabilities_one_valid)

print('Fraction:', best_repeat/10)
print('F1:', best_f1)
print('ROC AUC:', best_roc_auc)

Fraction: 0.4
F1: 0.6174055829228243
ROC AUC: 0.8293416190796103


As we can see the changes in sampling worsen the scores we are looking for. 

## Second model

Our second model is LogisticRegression. Here we do a similar test where we run the LogisticRegression several times with different configurations and sampling rates. 


In [150]:
model = LogisticRegression(random_state=12345, solver='liblinear')
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
probabilities_valid = model.predict_proba(features_test)
probabilities_one_valid = probabilities_valid[:, 1]
print('ROC AUC:', roc_auc_score(target_valid, probabilities_one_valid))

print(classification_report(target_valid, predicted_valid))

ROC AUC: 0.5470809539805173
              precision    recall  f1-score   support

           0       0.84      0.91      0.87       650
           1       0.65      0.49      0.56       229

    accuracy                           0.80       879
   macro avg       0.74      0.70      0.72       879
weighted avg       0.79      0.80      0.79       879



In [151]:
result = confusion_matrix(target_valid, predicted_valid)
print(result)

[[589  61]
 [116 113]]


In [152]:
model = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
probabilities_valid = model.predict_proba(features_test)
probabilities_one_valid = probabilities_valid[:, 1]
print('ROC AUC:', roc_auc_score(target_valid, probabilities_one_valid))

print(classification_report(target_valid, predicted_valid))

ROC AUC: 0.5461605643265033
              precision    recall  f1-score   support

           0       0.90      0.73      0.80       650
           1       0.50      0.77      0.60       229

    accuracy                           0.74       879
   macro avg       0.70      0.75      0.70       879
weighted avg       0.79      0.74      0.75       879



In [153]:
model = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')

best_f1 = 0
best_repeat = 0
best_report = 0
for repeat in range(1,51):
    features_upsampled, target_upsampled = upsample(
        features_train, target_train, repeat
    )
    
    model.fit(features_upsampled, target_upsampled)
    predicted_valid = model.predict(features_test)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_test)
    probabilities_one_valid = probabilities_valid[:, 1]

    if f1 > best_f1:
        best_f1 = f1
        best_repeat = repeat
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
        best_report = classification_report(target_valid, predicted_valid)

print(f"Repeat:{best_repeat} F1:{best_f1}")
print(best_report)
print(f"AUC-ROC: {auc_roc}")

Repeat:15 F1:0.3177257525083612
              precision    recall  f1-score   support

           0       0.74      0.58      0.65       650
           1       0.26      0.41      0.32       229

    accuracy                           0.54       879
   macro avg       0.50      0.50      0.48       879
weighted avg       0.61      0.54      0.56       879

AUC-ROC: 0.546133691635875


In [154]:
best_f1 = 0
best_repeat = 0
best_report = 0
auc_roc = 0
for repeat in range(1,11):
    features_downsampled, target_downsampled = downsample(
        features_train, target_train, (repeat/10)
    )

    model.fit(features_downsampled, target_downsampled)
    predicted_valid = model.predict(features_test)
    f1 = f1_score(target_valid, predicted_valid)
    probabilities_valid = model.predict_proba(features_test)
    probabilities_one_valid = probabilities_valid[:, 1]

    if f1 > best_f1:
        best_f1 = f1
        best_repeat = repeat
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
        best_report = classification_report(target_valid, predicted_valid)

print(f"Repeat:{best_repeat/10} F1:{best_f1}")
print(best_report)
print(f"AUC-ROC: {auc_roc}")

Repeat:0.1 F1:0.3293718166383701
              precision    recall  f1-score   support

           0       0.75      0.60      0.66       650
           1       0.27      0.42      0.33       229

    accuracy                           0.55       879
   macro avg       0.51      0.51      0.50       879
weighted avg       0.62      0.55      0.58       879

AUC-ROC: 0.5468928451461201


Unfortunately it seems like LogisticRegression algorithm is not optimal for our problem set. 

## RandomForestClassifier

Finally we will look at the RandomForestClassifier algo. Again we will run through several tests to figure out the optimal configuration and whether changing the sampling rate has an effect. 


In [155]:
best_depth = 0
best_est = 0
best_f1 = 0
best_roc_auc = 0
print("Depth:", end=' ')
for depth in range(1, 50):
    print(".", end='')
    for n_est in range(10, 111, 10):
        model = RandomForestClassifier(random_state=12345, n_estimators=n_est, max_depth=depth)
        model.fit(features_train, target_train)
        predicted_valid = model.predict(features_valid)
        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]

        f1 = f1_score(target_valid, predicted_valid)
        if best_f1 < f1:
            best_f1 = f1
            best_depth = depth
            best_est = n_est
            best_roc_auc = roc_auc_score(target_valid, probabilities_one_valid)
print()
print(f"Best n_est = {best_est}  max_depth = {best_depth} f1 score {best_f1} roc_auc {best_roc_auc}")

Depth: .................................................
Best n_est = 40  max_depth = 15 f1 score 0.5766233766233766 roc_auc 0.841891165602956


In [156]:
best_depth = 0
best_est = 0
best_f1 = 0
best_roc_auc = 0
print("Depth:", end=' ')
for depth in range(1, 10):
    print(".", end='')
    for n_est in range(10, 111, 10):
        model = RandomForestClassifier(random_state=12345, n_estimators=n_est, max_depth=depth, class_weight='balanced')
        model.fit(features_train, target_train)
        predicted_valid = model.predict(features_valid)
        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]

        f1 = f1_score(target_valid, predicted_valid)
        if best_f1 < f1:
            best_f1 = f1
            best_depth = depth
            best_est = n_est
            best_roc_auc = roc_auc_score(target_valid, probabilities_one_valid)
print()
print(f"Best n_est = {best_est}  max_depth = {best_depth} f1 score {best_f1} roc_auc {best_roc_auc}")

Depth: .........
Best n_est = 40  max_depth = 9 f1 score 0.6544789762340036 roc_auc 0.8482096069868995


In [162]:
best_depth = 9
best_est = 0
best_f1 = 0
best_samples = 0
best_leaf = 0
best_roc_auc = 0
print("n_est:", end=' ')
for n_est in range(10, 111, 10):
    print(".", end='')
    for min_samples in range(10, 111, 10):
        for min_sample_leaf in range(1, 11):
            model = RandomForestClassifier(random_state=12345, n_estimators=n_est, max_depth=best_depth, min_samples_split=min_samples, min_samples_leaf=min_sample_leaf, class_weight='balanced')
            model.fit(features_train, target_train)
            predicted_valid = model.predict(features_valid)
            probabilities_valid = model.predict_proba(features_valid)
            probabilities_one_valid = probabilities_valid[:, 1]

            f1 = f1_score(target_valid, predicted_valid)
            if best_f1 < f1:
                best_f1 = f1
                # best_depth = depth
                best_est = n_est
                best_samples = min_samples
                best_leaf = min_sample_leaf
                best_roc_auc = roc_auc_score(target_valid, probabilities_one_valid)
print()
print(f"Best n_est = {best_est}  max_depth = {best_depth} f1 score {best_f1} min_samples {best_samples} min_sample_leaf {best_leaf} roc_auc {best_roc_auc}")

n_est: ...........
Best n_est = 70  max_depth = 9 f1 score 0.6548042704626335 min_samples 10 min_sample_leaf 5 roc_auc 0.854779979845482


In [158]:
best_n_est = 70
best_max_depth = 9
best_min_samples = 40
best_min_sample_leaf = 5

In [159]:
result = confusion_matrix(target_valid, predicted_valid)
print(result)

[[483 167]
 [ 49 180]]


In [163]:
best_score = 0
best_repeat = 0
best_report = 0
best_roc_auc = 0
print("Repeat: ", end='')
for repeat in range(1,11):
    print(".", end='')
    features_upsampled, target_upsampled = upsample(features_train, target_train, repeat)
    model = RandomForestClassifier(random_state=12345, n_estimators=best_n_est, max_depth=best_max_depth, min_samples_leaf=best_min_sample_leaf, min_samples_split=best_min_samples, class_weight='balanced')
    model.fit(features_upsampled, target_upsampled)
    predicted_valid = model.predict(features_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    f1 = f1_score(target_valid, predicted_valid)
    if best_score < f1:
        best_roc_auc = roc_auc_score(target_valid, probabilities_one_valid)
        best_score = f1
        best_repeat = repeat
        best_report = classification_report(target_valid, predicted_valid)

print()
print("Repeat:",best_repeat)
print(best_report)
print(f"AUC-ROC: {best_roc_auc}")

Repeat: ..........
Repeat: 9
              precision    recall  f1-score   support

           0       0.92      0.75      0.83       650
           1       0.53      0.81      0.64       229

    accuracy                           0.77       879
   macro avg       0.73      0.78      0.73       879
weighted avg       0.82      0.77      0.78       879

AUC-ROC: 0.8528317097749412


In [164]:
best_score = 0
best_repeat = 0
best_report = 0
best_roc_auc = 0
print("Repeat: ", end='')
for repeat in range(1,100):
    print(".", end='')
    features_downsampled, target_downsampled = downsample(features_train, target_train, repeat/100)
    model = RandomForestClassifier(random_state=12345, n_estimators=best_n_est, max_depth=best_max_depth, min_samples_leaf=best_min_sample_leaf, min_samples_split=best_min_samples, class_weight='balanced')
    model.fit(features_downsampled, target_downsampled)
    predicted_valid = model.predict(features_valid)
    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    f1 = f1_score(target_valid, predicted_valid)
    if best_score < f1:
        best_roc_auc = roc_auc_score(target_valid, probabilities_one_valid)
        best_score = f1
        best_repeat = repeat
        best_report = classification_report(target_valid, predicted_valid)

print()
print("Repeat:",best_repeat/100)
print(best_report)
print(f"AUC-ROC: {best_roc_auc}")

Repeat: ...................................................................................................
Repeat: 0.97
              precision    recall  f1-score   support

           0       0.91      0.77      0.83       650
           1       0.54      0.78      0.64       229

    accuracy                           0.77       879
   macro avg       0.72      0.77      0.74       879
weighted avg       0.81      0.77      0.78       879

AUC-ROC: 0.8546422573060128


Our best score here is the last model with a 0.97 downsample, n_est of 70, max_depth of 9, min_samples of 40, and 5 leafs. The end result is a f1 score of 0.77 and a AUC-ROC score of 0.85. This is the highest combo of scores of all of our tests.

# Conclusion

In this project I explored several machine learning algorithms to predict the churn of Interconnect’s customers. Predicting churn for Interconnect is important to increase revenue and also have the ability to message customers before they leave the service. This also improves their retention rate. 

The first step in this process is to load and clean up the data. First the csv data is loaded and joined on the customerId column. Next we clean up the data such as the date columns. These columns were in an object format and converting them to datetime format makes this data easier to work with. The category columns (ex 'MultipleLines', 'Partner', 'Dependents', etc) can be converted into true/false columns. At this point a new column can be created called ‘Churn’. The value (true/false) is based on whether the end date is null. A null end date indicates the customer has not left Interconnect. This will be the target column. Finally we can get dummies for several category columns with multiple categories. This transforms text based categories into numerical values.

Now the data is prepared, the next phase of the project can begin. First the target column ‘Churn’ is separated into a target and the rest is used as features. Several columns are also dropped ('customerID', 'Churn', 'BeginDate', 'EndDate') that do not provide us any value. Here the feature and target data is split into test, train, and validation sets. There are also two columns 'MonthlyCharges', 'TotalCharges' that are scaled. Scaling data allows us to flatten out any outliers and noise in our data. 

DecisionTreeClassifier is the first attempt at solving Interconnect’s churn prediction problem. First step is to find the optimal max depth for the tree. In this case the max depth is 6 and produces an f1 score of 0.79 and a AUC-ROC score of 0.82. This potentially can be improved on by upsampling or downsampling the data. Upsampling or downsampling alters the data by increasing the number of data points to create a stronger signal or removes data points to create less noise. Using the confusion matrix function there is an imbalance in classes and changing the sampling rate can improve the performance. For the DecisionTreeClassifier upsampling/downsampling only creates a small increase in performance. 

The second model is LogisticRegression. Here the same steps are performed. First the base line is established and produces an f1 score of 0.74 and AUC-ROC 0.54. Upsampling and downsampling does not improve the model and the base line results can be improved upon.

The third algorithm tried is the RandomForestClassifier algorithm. RandomForestClassifier has more hyperparameters to tune but follows a similar pattern to the previous algorithms. The hyperparameters that are being optimized are the n_estimators and max_depth. The f1 score is 0.57 and the AUC-ROC score 0.84. If the classes are balanced using the class_weight hyperparameter then the scores improve slightly with an f1 score of ​​0.65 roc_auc 0.84. Next the hyperparameters min_samples_split and min_samples_leaf can be tuned. Similar to the other algorithms, changing the sampling rate may improve the performance. Combining the hyperparameter tuning and downsampling produces good results. The end result is a f1 score of 0.77 and a AUC-ROC score of 0.85 which is the highest score. 
This project was challenging in that there are many ways to solve this problem and only by experimenting will the solution be discovered. The end result is that a solid solution has been discovered that Interconnect can use to identify what users are about to leave the service. 
